### PR TITLE
Frida Codeshare integration! 🎉

### PR DESCRIPTION
**Pull request by Fitblip**
This is my first pass at adding the codeshare integration with `frida-python`. Give it a gander and let me know what you think, it's functional but there are probably coding conventions/style items that need to change to adhere to the rest of the frida project's style, so I'm happy to get even pedantic feedback! :) 

Notable changes:
- Added the `-c`/`--codeshare` argument that takes in a project
URI/slug (ex. `fitblip/hello-world`), pulls down the source code from
codeshare.frida.re, checks the signature and either prompts the user to
accept it and run or reject it.
- I create and use `~/.frida/` as a storage location for the
`codeshare-truststore.json` file. We might consider moving the
.frida-history there, but I didn’t want to just do it and break
people’s current .frida-history without discussing it first.
- I have my string constants inline with the function that uses them,
if you have a different convention for using string literals I’m happy
to adhere, but I didn’t see one!
- I added `.idea/` to `.gitignore`, it's a directory created by PyCharm (my IDE of choice). We can remove it if you prefer not to have it though!
- I also fixed a single import (`platform`) which didn't seem to be used. 

Misc:
- Right now it doesn’t do any checking to see if someone has also
specified a user script, if they do we just clobber it. Would you
prefer I catch this situation as an invalid condition or combine the
two scripts together? If we concatenate them we might consider moving
the codeshare payload to the front of the script so that people can
write more library-esque snippets and extend them with custom logic
locally. Just a thought!
- I’m using `raw_input`/`input`, though I see you’re also doing some
console stuff to make completion and whatnot work, do you think that’ll
cause any issues? It doesn’t seem to in any of my testing.
- The current behavior is that if the user rejects a fingerprint we
still drop into a shell, do you think it’d make sense to
- Should we add some sort of version tracking/user agent? Stats might
be cool around how many times stuff has been downloaded and by what
versions. Do you do any metrics collection for Frida?
- Generally people can be somewhat upset when something reaches out to
the internet, would it make sense for us to warn that this
functionality does this? I’d hope that most people already know that,
but who knows lol.

You know how to get ahold of me if you have questions!

### CODE REVIEW

**Comment by oleavr** [comment]
w00t, awesome stuff!

> , checks the signature and either prompts the user to accept it

I guess we also need to show this signature on the web so users can compare?

> We might consider moving the .frida-history there

I think we could `os.path.exists()` on the old location and then `shutil.move()` it into the new location? We could just put a fat TODO there to remind us to remove it for the next major release, or some point in the future when all users are likely to be running a recent version.

> Right now it doesn’t do any checking to see if someone has also
> specified a user script, if they do we just clobber it.

I think it would be useful to support this. Definitely put the user script last, though, so it can use functions defined by the snippit.

> I’m using raw_input/input

This should be fine until the UI thread has grabbed `stdin`.

> The current behavior is that if the user rejects a fingerprint we
> still drop into a shell

Would probably be clearer to exit in case the user script depends on things from the snippit.

> Should we add some sort of version tracking/user agent

Would probably be nice to include the Frida version and OS. We don't currently collect anything, so this being an obvious online feature I think it makes sense to do it here.

> Generally people can be somewhat upset when something reaches out to
> the internet, would it make sense for us to warn that this
> functionality does this? I’d hope that most people already know that,
> but who knows lol.

No I don't think it's worth the trouble. They already got the ID from a website so that should be obvious enough. :)

File: src/frida/repl.py [hunk 1]
```diff
@@ -423,6 +431,57 @@ def _create_repl_script(self):
...
+                except Exception as e:
+                    self._print("Got an unhandled exception while trying to retrieve {} - {}".format(self._codeshare_uri, e))
+
+                if response_json:
+                    trusted_signature = trust_store.get(self._codeshare_uri, '')
+                    fingerprint = hashlib.sha256(response_json['source'].encode('utf-8')).hexdigest()
+                    if fingerprint == trusted_signature:
+                        user_script = response_json['source']
+                    else:
+                        self._print("""Hello! This is the first time you're running this particular snippit, or the snippit's source code has changed.
```
Comment (oleavr) [inline src/frida/repl.py]: Is it a `snippit` or a `snippet`? I have no idea, you're the native speaker, so definitely your call! :)

File: src/frida/repl.py [hunk 2]
```diff
@@ -423,6 +431,57 @@ def _create_repl_script(self):
...
+                        ))
+
+                        while True:
+                            input_string = "Are you sure you'd like to trust this project? [y/N] "
+                            try:
+                                response = raw_input(input_string)
+                            except NameError:
+                                response = input(input_string)
+
+                            if response.lower() in 'no' or response == '':
```
Comment (oleavr) [inline src/frida/repl.py]: This means that both `no`, `n`, and `o` are valid choices. `o` seems a bit weird. Perhaps we want something like:

```py
if response.lower() in ('no', 'n') or response == '':
```

File: src/frida/repl.py [hunk 3]
```diff
@@ -423,6 +431,57 @@ def _create_repl_script(self):
...
+                                self._print("Dropping into a normal REPL shell!")
+                                break
+
+                            if response.lower() in 'yes':
```
Comment (oleavr) [inline src/frida/repl.py]: Like above, perhaps:
```py
if response.lower() in ('yes', 'y'):
```
(As `e` and `s` aren't that intuitive.)

File: src/frida/repl.py [hunk 4]
```diff
@@ -423,6 +431,57 @@ def _create_repl_script(self):
                 with codecs.open(self._user_script, 'rb', 'utf-8') as f:
                     user_script = f.read().rstrip("\r\n") + "\n\n// Frida REPL script:\n"
 
+            if self._codeshare_uri is not None:
+                trust_store = self._get_or_create_truststore()
+                project_url = 'https://codeshare.frida.re/api/project/{}/'.format(self._codeshare_uri)
```
Comment (oleavr) [inline src/frida/repl.py]: We should stick to double quotation marks and only use single quotation marks for symbols/atoms (enum values). In hindsight this is probably too weird and something we might want to move away from at some point (in one big swoop), but in the meantime it's good to be consistent.

File: src/frida/repl.py [hunk 5]
```diff
@@ -423,6 +431,57 @@ def _create_repl_script(self):
...
+                response_json = None
+                try:
+                    response = urlopen(project_url)
+                    response_content = response.read()
+                    response_json = json.loads(response_content)
+                except Exception as e:
+                    self._print("Got an unhandled exception while trying to retrieve {} - {}".format(self._codeshare_uri, e))
+
+                if response_json:
+                    trusted_signature = trust_store.get(self._codeshare_uri, '')
```
Comment (oleavr) [inline src/frida/repl.py]: Quotation marks here as well.

File: src/frida/repl.py [hunk 6]
```diff
@@ -423,6 +431,57 @@ def _create_repl_script(self):
...
+                    fingerprint = hashlib.sha256(response_json['source'].encode('utf-8')).hexdigest()
+                    if fingerprint == trusted_signature:
+                        user_script = response_json['source']
```
Comment (oleavr) [inline src/frida/repl.py]: (Just to be clear: quotation marks on these three lines are correct.)

File: src/frida/repl.py [hunk 7]
```diff
@@ -443,6 +502,42 @@ def _create_repl_script(self):
...
+
+            with open(codeshare_trust_store, 'w') as f:
+                f.write(json.dumps(trust_store))
+
+        def _get_or_create_truststore(self):
+            '''
+            Look for a trust store located at ~/.frida/codeshare-truststore.json, which holds known fingerprints of scripts previously
+            run before, prompting if the fingerprint changes (due to a code change pushed to codeshare).
+            :return:
+            '''
```
Comment (oleavr) [inline src/frida/repl.py]: Quotation marks should be adjusted here.

File: src/frida/repl.py [hunk 8]
```diff
@@ -443,6 +502,42 @@ def _create_repl_script(self):
...
+            config_dir = os.path.join(os.path.expanduser('~'), '.frida')
```
Comment (oleavr) [inline src/frida/repl.py]: Quotation marks need adjusting.

File: src/frida/repl.py [hunk 9]
```diff
@@ -19,6 +19,11 @@ def main():
     import sys
     import threading
```
Comment (Fitblip) [inline src/frida/repl.py]: Done!

File: src/frida/repl.py [hunk 10]
```diff
@@ -37,6 +42,8 @@ def __init__(self):
         def _add_options(self, parser):
             parser.add_option("-l", "--load", help="load SCRIPT", metavar="SCRIPT",
                 type='string', action='store', dest="user_script", default=None)
+            parser.add_option("-c", "--codeshare", help="load CODESHARE_URI", metavar="CODESHARE_URI",
+                type='string', action='store', dest="codeshare_repo", default=None)
```
Comment (Fitblip) [inline src/frida/repl.py]: It should indeed :-P

File: src/frida/repl.py [hunk 11]
```diff
@@ -423,6 +431,57 @@ def _create_repl_script(self):
                 with codecs.open(self._user_script, 'rb', 'utf-8') as f:
                     user_script = f.read().rstrip("\r\n") + "\n\n// Frida REPL script:\n"
 
+            if self._codeshare_uri is not None:
```
Comment (Fitblip) [inline src/frida/repl.py]: I think it makes sense for `%reload` to reload it tbh, it's what I'd more or less expect it to do, but that just might be me. What do you think? 

> but if we do we need to make sure that it works at that point.

I'm not sure I follow what you're saying here, can you say that a different way? Are you saying that the UI thread could potentially be active by the time this is running?

File: src/frida/repl.py (hunk 4 above)
Comment (Fitblip) [inline src/frida/repl.py]: Agreed! That's normally what I try to do actually, but I do occasionally mix them up so your point might be valid :-P. Either way I'll make sure to pay better attention for that in future commits!

File: src/frida/repl.py (hunk 1 above)
Comment (Fitblip) [inline src/frida/repl.py]: Oh man! I guess I'm barely a native speaker :-P, it's definitely snippet, I just can't spell.

File: src/frida/repl.py (hunk 2 above)
Comment (Fitblip) [inline src/frida/repl.py]: That's fair, I figured this was a more succinct way to do it, but you're right that it'll accept a few keys as y/n, I just figured no-one would ever hit those corner cases :-P. Happy to change it to be more explicit though! :)

File: src/frida/repl.py (hunk 7 above)
Comment (Fitblip) [inline src/frida/repl.py]: I also notice you don't have a lot of comments, I'm happy to remove this if you'd like as it's not exactly rocket science to figure out what this does.

File: src/frida/repl.py (hunk 8 above)
Comment (Fitblip) [inline src/frida/repl.py]: Done!

**Comment by Fitblip** [comment]
> I guess we also need to show this signature on the web so users can compare?

Ah good point, I'll add that in right now.

**Comment by oleavr** [comment]
Awesome! Totally agree on the migration, not a big fan of that kind of logic either.

Alright, let's get this released!

**Comment by Fitblip** [comment]
Excellent! Thanks for the merge :)

**Pull request merged**